src/deriveNativeScriptHash.c

```diff
@@ -0,0 +1,341 @@
+		ADDRESS_KEY_HASH_LENGTH,
+		&deriveScriptHash_display_ui_callback
+	);
+}
+
+void deriveNativeScriptHash_handlePubkey(read_view_t* view)
+{
+	uint8_t pubkeyType = parse_u1be(view);
+	TRACE("pubkey type = %u", pubkeyType);
+
+	switch(pubkeyType) {
+#	define  CASE(TYPE, HANDLER) case TYPE: HANDLER(view); break;
+		CASE(PUBKEY_DEVICE_OWNED, deriveNativeScriptHash_handleDeviceOwnedPubkey);
+		CASE(PUBKEY_THIRD_PARTY, deriveNativeScriptHash_handleThirdPartyPubkey);
+		// any other value for the pubkey type is invalid
+		default: THROW(ERR_INVALID_DATA);
+#	undef   CASE
+	}
+}
+
+#define SPLIT_UINT64(n) (uint32_t) (n >> 32), (uint32_t) n
+
+void deriveNativeScriptHash_handleInvalidBefore(read_view_t* view)
+{
+	ctx->nativeScript.invalidBefore.timelock = parse_u8be(view);
+
+	VALIDATE(view_remainingSize(view) == 0, ERR_INVALID_DATA);
+	// PRINTF doesn't handle uint64_t so we split it into two uint32_t
+	TRACE("invalid_before timelock = 0x%x%x", SPLIT_UINT64(ctx->nativeScript.invalidBefore.timelock));
+
+	ui_displayUint64Screen(
+		"Script - invalid before",
+		ctx->nativeScript.invalidBefore.timelock,
+		&deriveScriptHash_display_ui_callback
+	);
+}
+
+void deriveNativeScriptHash_handleInvalidHereafter(read_view_t* view)
+{
+	ctx->nativeScript.invalidHereafter.timelock = parse_u8be(view);
+
+	VALIDATE(view_remainingSize(view) == 0, ERR_INVALID_DATA);
+	TRACE("invalid_hereafter timelock = 0x%x%x", SPLIT_UINT64(ctx->nativeScript.invalidHereafter.timelock));
+
+	ui_displayUint64Screen(
+		"Script - invalid hereafter",
+		ctx->nativeScript.invalidHereafter.timelock,
+		&deriveScriptHash_display_ui_callback
+	);
+}
+
+#undef SPLIT_UINT64
+
+void deriveNativeScriptHash_handleSimpleScript(read_view_t* view)
+{
+	TRACE_BUFFER(view->begin, (size_t) (view->end - view->begin));
+
+	VALIDATE(expectsRequest(), ERR_INVALID_STATE);
+
+	uint8_t nativeScriptType = parse_u1be(view);
+	TRACE("native simple script type = %u", nativeScriptType);
+
+	// parse data
+	switch(nativeScriptType) {
+#	define  CASE(TYPE, HANDLER) case TYPE: HANDLER(view); break;
+		CASE(NATIVE_SCRIPT_PUBKEY, deriveNativeScriptHash_handlePubkey);
+		CASE(NATIVE_SCRIPT_INVALID_BEFORE, deriveNativeScriptHash_handleInvalidBefore);
+		CASE(NATIVE_SCRIPT_INVALID_HEREAFTER, deriveNativeScriptHash_handleInvalidHereafter);
+		default: THROW(ERR_INVALID_DATA);
+#	undef   CASE
+	}
+	scriptFinished();
+}
+
+// Complex script finish
+
+void deriveNativeScriptHash_handleComplexScriptFinish(read_view_t* view)
+{
+	// expect no data
+	VALIDATE(read_view_remaining_size(view) == 0, ERR_INVALID_DATA);
+
+	// if the number of remaining scripts for the current complex script is not 0
+	// then this request is invalid in the current context
+	VALIDATE(ctx->remainingScripts[ctx->level] == 0, ERR_INVALID_STATE);
+
+	switch(ctx->type[ctx->level]) {
+#	define  CASE(TYPE, HEADER) case TYPE: ui_displayPaginatedText(HEADER, "Finished", &deriveScriptHash_display_ui_callback); break;
```

Comment:
Does this mean that also other references in the file to `&deriveScriptHash_display_ui_callback` should be changed to `deriveScriptHash_display_ui_callback`?

Also if I may ask, what's the difference? Because looking at the code I am a little bit confused. For example calling the `ui_displayUint64Screen` accepts `ui_callback_fn_t callback`:
```c
void ui_displayUint64Screen(
        const char* firstLine,
        uint64_t value,
        ui_callback_fn_t callback
)
{
	char valueStr[30];
	str_formatUint64(value, valueStr, SIZEOF(valueStr));

	ui_displayPaginatedText(
	        firstLine,
	        valueStr,
	        callback
	);
}
```
but the `ui_displayPaginatedText` has this signature:
```c
void ui_displayPaginatedText(
        const char* headerStr,
        const char* bodyStr,
        ui_callback_fn_t* callback)
```
namely the `callback` here is a pointer, but in the `ui_displayUint64Screen` the original `callback` is passed as it is. So what's going on here?

To me it seems like `ui_callback_fn_t*` and `ui_callback_fn_t` are the same, but I am probably missing something

Reply:
In C, the function name is also a pointer to the function. It is common to use it in that way, e.g. when you have a function pointer f, you usually call the function `f()`, not `(*f)()`, though both work. Similarly, you pass a function g() as either `g` or `&g` as parameter, and our codebase uses the first way, so we want to be consistent.
https://stackoverflow.com/questions/12152167/why-is-using-the-function-name-as-a-function-pointer-equivalent-to-applying-the